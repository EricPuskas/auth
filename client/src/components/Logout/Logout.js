import React from "react";
import PropTypes from "prop-types";

const Logout = ({ logoutUser }) => {
  return (
    <button
      data-test="logout-btn"
      onClick={logoutUser}
      className="btn btn-primary"
    >
      <i className="fas fa-sign-out-alt" /> Logout
    </button>
  );
};

Logout.propTypes = {
  logoutUser: PropTypes.func
};

export default Logout;
