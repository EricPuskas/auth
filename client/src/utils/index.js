import checkPropTypes from "check-prop-types";
import { applyMiddleware, createStore } from "redux";
import rootReducer from "../reducers";
import { middleware } from "../store";

export const isEmpty = value =>
  value === undefined ||
  value === null ||
  (typeof value === "object" && Object.keys(value).length === 0) ||
  (typeof value === "string" && value.trim().length === 0);

export const findByAttribute = (component, attr) =>
  component.find(`[data-test="${attr}"]`);

export const checkProps = (component, expectedProps) => {
  return checkPropTypes(
    component.propTypes,
    expectedProps,
    "props",
    component.name
  );
};

export const isSecure = password => {
  let error_message = "";
  let lowerRegex = /[a-z]/;
  let upperRegex = /[A-Z]/;
  let numberRegex = /[0-9]/;
  let specialRegex = /[$@$!%*?&]/;

  if (isEmpty(password)) {
    error_message = "Password field is required.";
  }
  if (!password.match(lowerRegex))
    error_message = "Password must contain at least one lowercase letter.";
  if (!password.match(upperRegex))
    error_message = "Password must contain at least one uppercase letter.";
  if (!password.match(numberRegex))
    error_message = "Password must contain at least one number.";
  if (!password.match(specialRegex))
    error_message = "Password must contain at least a special character";
  if (password.length < 8 || password.length > 30) {
    error_message = "Password must be between 8 and 30 characters.";
  }
  return error_message;
};

export const testStore = initState => {
  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(rootReducer, initState);
};
