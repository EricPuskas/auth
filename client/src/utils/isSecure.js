import isEmpty from "./isEmpty";

const isSecure = password => {
  let error_message = "";
  let lowerRegex = /[a-z]/;
  let upperRegex = /[A-Z]/;
  let numberRegex = /[0-9]/;
  let specialRegex = /[$@$!%*?&]/;

  if (isEmpty(password)) {
    error_message = "Password field is required.";
  }
  if (!password.match(lowerRegex))
    error_message = "Password must contain at least one lowercase letter.";
  if (!password.match(upperRegex))
    error_message = "Password must contain at least one uppercase letter.";
  if (!password.match(numberRegex))
    error_message = "Password must contain at least one number.";
  if (!password.match(specialRegex))
    error_message = "Password must contain at least a special character";
  if (password.length < 8 || password.length > 30) {
    error_message = "Password must be between 8 and 30 characters.";
  }
  return error_message;
};

export default isSecure;
