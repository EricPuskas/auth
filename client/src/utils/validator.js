import { isSecure } from "./index";

const Validator = {
  login: data => {
    const errors = {};
    if (!data.password) {
      errors.password = "Password field is required.";
    } else {
      errors.password = isSecure(data.password);
    }
    if (!data.email) {
      errors.email = "Email field is required.";
    } else if (!/^[^@]+@[^@]+\.[^@]+$/.test(data.email)) {
      errors.email = "Please enter a valid email address (example@email.com).";
    }
    return errors;
  },
  register: data => {
    const errors = {};
    if (!data.password) {
      errors.password = "Password field is required.";
    } else {
      errors.password = isSecure(data.password);
    }
    if (!data.email) {
      errors.email = "Email field is required.";
    } else if (!/^[^@]+@[^@]+\.[^@]+$/.test(data.email)) {
      errors.email = "Please enter a valid email address (example@email.com).";
    }
    if (!data.firstName) {
      errors.firstName = "First Name field is required.";
    }
    if (!data.lastName) {
      errors.lastName = "Last Name field is required.";
    }
    return errors;
  }
};

export default Validator;
