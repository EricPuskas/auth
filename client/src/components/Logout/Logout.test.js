import React from "react";
import { shallow } from "enzyme";
import { findByAttribute, checkProps } from "../../utils";
import Logout from "./Logout";

describe("Logout Component", () => {
  describe("Checking PropTypes", () => {
    it("Should not throw a warning", () => {
      const expectedProps = {
        logoutUser: jest.fn()
      };
      const propsError = checkProps(Logout, expectedProps);
      expect(propsError).toBeUndefined();
    });
  });
});

describe("Renders without crashing", () => {
  let component;
  let mockFn;
  beforeEach(() => {
    mockFn = jest.fn();
    const props = {
      logoutUser: mockFn
    };
    component = shallow(<Logout {...props} />);
  });
  it("Should pass snapshot test", () => {
    expect(component.html()).toMatchSnapshot();
  });

  it("Should render a button", () => {
    const button = findByAttribute(component, "logout-btn");
    expect(button.length).toBe(1);
  });

  it("Should emit a callback on click event", () => {
    const button = findByAttribute(component, "logout-btn");
    button.simulate("click");
    const callback = mockFn.mock.calls.length;
    expect(callback).toBe(1);
  });
});
