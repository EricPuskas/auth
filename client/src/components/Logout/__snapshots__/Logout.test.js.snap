// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crashing Should pass snapshot test 1`] = `"<button data-test=\\"logout-btn\\" class=\\"btn btn-primary\\"><i class=\\"fas fa-sign-out-alt\\"></i> Logout</button>"`;
